### DISCUSSION TITLE
Alpine + HTMX

### DISCUSSION THREAD

**Discussion by SirMetathyst** [alpinejs/alpine#2605] [1. Help]
Will alpine work if I fetch a html partial from the backend and it contains an alpine component?

Example:
```html
<div hx-target="this" hx-swap="outerHTML">
    <p>comment text...</p>
    <button hx-get="/comment/1/edit" class="btn btn-primary">
    <!-- Html with alpine component replacing div... textarea form with alpine word/character count? -->
    </button>
</div>
```

**Comment by mshamma** [comment]
Not your specific use case, but I use a wrapper method for 'bridging' Alpine and HTMX.  

```
document.addEventListener('alpine:init', () => {
    initWrapper(null);
});

// could be any htmx event here
document.body.addEventListener('htmx:beforeSwap', function (evt) {
    initWrapper(evt);
});

function initWrapper(evt) {
        // I use Alpine.store() as one global data facade for my site.
        Alpine.store('darkmode', {
            on: false,
 
            toggle() {
                this.on = ! this.on
            }
            // add more stuff here
        })
}
```

**Reply by skttl** [reply]
Hi @mshamma 

I tried doing like you, and ultimately things work. But I do get a slight flickr of a non-initialized Alpine component, and a bunch of warnings and errors in my console, like Alpine Expression errors saying that eg. my data component name doesn't exist.

If I send a partial like this
```
<div x-data="dropdown">
    <button x-on:click="toggle">...</button>

    <div x-show="open">...</div>
</div>

<script>
    document.addEventListener('alpine:init', () => {
        initWrapper(null);
    });

    // could be any htmx event here
    document.body.addEventListener('htmx:beforeSwap', function (evt) {
        initWrapper(evt);
    });

    function initWrapper(evt) {
        Alpine.data('dropdown', () => ({
            open: false,

            toggle() {
                this.open = !this.open
            }
        }))
    };
</script>
```

I get warnings in the console like this:

![image](https://github.com/alpinejs/alpine/assets/3726467/88f6b0ae-39c0-4dd3-a6ba-d43e1b999803)

Ultimately the component works - if I click the button, the div shows. I think HTMX might be adding the html, then running the script.

Do you have the same experience?

**Reply by ekwoka** [reply]
You need to run the scripts before the html is added.

And make sure you aren't calling Alpine.start twice.

You will likely need to add a hook to htmx to separate the scripts from the dom and add them in two steps.

**Reply by heitorado** [reply]
Hi! Sorry for jumping into the conversation but I'm facing the same issue. Does anyone have any suggestion of a good way to send partials with Alpine code to HTMX swaps and ensuring that the partial works without any errors?

> You need to run the scripts before the html is added.

@ekwoka how can I enforce that?

> You will likely need to add a hook to htmx to separate the scripts from the dom and add them in two steps.

@ekwoka what do you mean with "add a hook to htmx"? Do you have any code example to share?
Adding the script first and then the HTML seems a bit clunky IMHO. And you would need two separate endpoints. This can get out of hand fast and doesn't seem like a pleasant front end devex. I'm searching for better ways to do that but I'm not finding anything.

**Reply by ekwoka** [reply]
You use HTMXs events like before swap, to extract the script tags and add them first, before letting htmx add the whole template.

**Reply by heitorado** [reply]
I see. Would you have a code example for achieving that? I'm familiar with `beforeSwap` but I still don't understand how you would extract a part of the content being swapped to add it first.

**Reply by ekwoka** [reply]
The `xhr` should have the content returned from the server on it, so you can then parse and query it yourself, even replacing the whole swap logic if you want or just extract the scripts and create new ones.

**Reply by heitorado** [reply]
Here's my attempt of doing something that would be generic and work across all my application:
```js
document.body.addEventListener('htmx:beforeSwap', function (evt) {
    dp = new DOMParser()

    fragmentToSwap = dp.parseFromString(evt.detail.xhr.response, "text/html")
    fragmentToSwap.querySelectorAll('script').forEach(script => {
        // Clone the script node, append to the end of the DOM, and remove the script from the fragment that will be swapped
        const renderAheadScript = script.cloneNode(true);
        document.body.appendChild(renderAheadScript);
        script.remove();
    })
});

// However, we have a problem here. I cannot assign anything to evt.detail.xhr.response because that's a read-only property
```

MDN Docs: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest

If I can't change the original response, I'm going to end up with two scripts. I don't want to change anything else in the original HTMX swap behavior. Just want to ensure that the scripts with alpine stuff are executed first so I don't end up with console errors of functions and state variables not being defined.

**Reply by SimoTod** [reply]
I'm not sure trying to manipulate the response is a good idea. Can you try something as simple as

```javascript
    <script>
      document.addEventListener('htmx:beforeSwap', function (event) {
        Alpine.deferMutations()
      })
      document.addEventListener('htmx:afterSettle', function (event) {
        Alpine.flushAndStopDeferringMutations()
      })
    </script>
```

I tried on an online playground an it worked for me. The code just stops Alpine from trying to initialise anything before the swap, and it restart it after the dom is settled (there's also a afterSwap event but I think it fires too early).

**Reply by ekwoka** [reply]
@SimoTod This doesn't handle the issue of script tags themselves being marked as started by the parsers. The script tags won't evaluate. 

Or does HTMX already handle that?

**Reply by SimoTod** [reply]
@ekwoka I think HTMX handles it
See https://lassebomh.github.io/htmx-playground/?url=https%253A%252F%252Fgist.githubusercontent.com%252FSimoTod%252F2c82791bb30333061af7de16f154a52a%252Fraw%252F716f4457f370f6b41f157969151f8ca21a9aeb45%252Fhtmx%25252520test

**Reply by SimoTod** [reply]
It uses https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L570

**Reply by ekwoka** [reply]
Right on
